### PR TITLE
Allow conditional delete in List access checker

### PR DESCRIPTION
**Pull request by vivekmittal07**
## Description of what I changed
Added support for conditional delete in List access checker

Issue https://github.com/google/fhir-access-proxy/issues/43

## E2E test

TESTED: Verified the delete workflows locally for a patient and observations for patients in the access list.

## Checklist: I completed these to help reviewers :)

- [x] I have read and will follow the [review process](https://github.com/GoogleCloudPlatform/openmrs-fhir-analytics/blob/master/doc/review_process.md).
- [x] I am familiar with Google Style Guides for the language I have coded in.

  No? Please take some time and review [Java](https://google.github.io/styleguide/javaguide.html) and [Python](https://google.github.io/styleguide/pyguide.html) style guides.

- [x] My IDE is configured to follow the Google [**code styles**](https://google.github.io/styleguide/).

  No? Unsure? -> [configure your IDE](https://github.com/google/google-java-format).

- [x] I have **added tests** to cover my changes. (If you refactored existing code that was well tested you do not have to add tests)
- [x] I ran `mvn clean package` right before creating this pull request and added all formatting changes to my commit.
- [x] All new and existing **tests passed**.
- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: plugins/src/main/java/com/google/fhir/gateway/plugin/ListAccessChecker.java [hunk 1]
```diff
@@ -227,6 +228,22 @@ private AccessDecision processPatch(RequestDetailsReader requestDetails) throws
     return checkNonPatientAccessInUpdate(requestDetails, RequestTypeEnum.PATCH);
   }
 
+  private AccessDecision processDelete(RequestDetailsReader requestDetails) {
+    // We don't support deletion of List resource used as an access list for a user.
+    if (FhirUtil.isSameResourceType(requestDetails.getResourceName(), ResourceType.List)) {
+      if (patientListId.equals(FhirUtil.getIdOrNull(requestDetails))) {
```
Comment (bashir2) [inline plugins/src/main/java/com/google/fhir/gateway/plugin/ListAccessChecker.java:234]: Hmm, shouldn't this be _not_ `equals`?

File: plugins/src/test/java/com/google/fhir/gateway/plugin/ListAccessCheckerTest.java [hunk 2]
```diff
@@ -17,8 +17,7 @@
 
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.equalTo;
-import static org.mockito.Mockito.doReturn;
-import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.*;
```
Comment (bashir2) [inline plugins/src/test/java/com/google/fhir/gateway/plugin/ListAccessCheckerTest.java:20]: Please avoid using wildcard imports ([style guide](https://google.github.io/styleguide/javaguide.html#s3.3.1-wildcard-imports)).

File: plugins/src/main/java/com/google/fhir/gateway/plugin/ListAccessChecker.java [hunk 3]
```diff
@@ -227,6 +228,22 @@ private AccessDecision processPatch(RequestDetailsReader requestDetails) throws
     return checkNonPatientAccessInUpdate(requestDetails, RequestTypeEnum.PATCH);
   }
 
+  private AccessDecision processDelete(RequestDetailsReader requestDetails) {
+    // We don't support deletion of List resource used as an access list for a user.
```
Comment (bashir2) [inline plugins/src/main/java/com/google/fhir/gateway/plugin/ListAccessChecker.java:232]: In general, I think we should not let deletion of any `List` resource without a patient parameter, should we?

**Review by bashir2 — COMMENTED** [review]
Thanks @vivekmittal07 for the change. I only have a minor comment about `List` deletion.

File: plugins/src/main/java/com/google/fhir/gateway/plugin/ListAccessChecker.java (hunk 3 above)
Comment (vivekmittal07) [reply plugins/src/main/java/com/google/fhir/gateway/plugin/ListAccessChecker.java:232]: yes right.
It is same as deletion of any other resource. Fixed this.

File: plugins/src/test/java/com/google/fhir/gateway/plugin/ListAccessCheckerTest.java (hunk 2 above)
Comment (vivekmittal07) [reply plugins/src/test/java/com/google/fhir/gateway/plugin/ListAccessCheckerTest.java]: done

File: plugins/src/main/java/com/google/fhir/gateway/plugin/ListAccessChecker.java (hunk 1 above)
Comment (vivekmittal07) [reply plugins/src/main/java/com/google/fhir/gateway/plugin/ListAccessChecker.java]: Thanks for catching this. Intent was this  only but made a mistake below.
Fixed this and also added the required test case.

File: plugins/src/main/java/com/google/fhir/gateway/plugin/ListAccessChecker.java [hunk 4]
```diff
@@ -171,8 +171,9 @@ public AccessDecision checkAccess(RequestDetailsReader requestDetails) {
           return processPut(requestDetails);
         case PATCH:
           return processPatch(requestDetails);
+        case DELETE:
+          return processDelete(requestDetails);
```
Comment (bashir2) [inline plugins/src/main/java/com/google/fhir/gateway/plugin/ListAccessChecker.java:175]: Sorry, one thing that I remembered after sending the previous review is handling DELETE in a Bundles which we should address as part of #43 . This also applies to #72. If you want to have a separate PR for that, it is fine, but please do so before closing #43.

**Review by bashir2 — APPROVED** [review]
Thanks for the changes. The main remaining part is handling DELETE in Bundles which can be done in a separate PR (for both access-checkers).

**Review by vivekmittal07 — COMMENTED** [review]
I will handle the Delete in FHIR bundles in a different PR.
Ptal at the PR again.

**Pull request merged**